timeline/src/Data/Timeline/Internal.hs

```diff
@@ -0,0 +1,336 @@
+    unpackOverlapGroup,
+  )
+where
+
+import Data.Foldable.WithIndex (FoldableWithIndex (..))
+import Data.Functor.Contravariant (Contravariant, contramap)
+import Data.Functor.WithIndex (FunctorWithIndex (..))
+import Data.List (intercalate, sortOn)
+import Data.List.NonEmpty (NonEmpty ((:|)), nonEmpty)
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.Map.Merge.Strict qualified as Map
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.Maybe (mapMaybe, maybeToList)
+import Data.Semigroup.Foldable.Class (fold1)
+import Data.Set (Set)
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Time
+  ( UTCTime (..),
+    diffTimeToPicoseconds,
+    picosecondsToDiffTime,
+  )
+import Data.Time.Calendar.OrdinalDate (fromOrdinalDate, toOrdinalDate)
+import Data.Traversable.WithIndex (TraversableWithIndex (..))
+import GHC.Generics (Generic)
+import GHC.Records (HasField (getField))
+import Language.Haskell.TH.Syntax qualified as TH (Lift (liftTyped))
+import Language.Haskell.TH.Syntax.Compat qualified as TH
+import Prelude
+
+-- | A unbounded discrete timeline for data type @a@. @'Timeline' a@ always has
+-- a value for any time, but the value can only change for a finite number of
+-- times.
+--
+-- * 'Functor', 'Foldable' and 'Traversable' instances are provided to traverse
+--   through the timeline;
+-- * 'FunctorWithIndex', 'Foldable' and 'TraversableWithIndex' instances are
+-- provided in case you need the current time range where each value holds
+-- * 'Applicative' instance can be used to merge multiple 'Timeline's together
+data Timeline a = Timeline
+  { -- | the value from negative infinity time to the first time in 'values'
+    initialValue :: a,
+    -- | changes are keyed by their "effective from" time, for easier lookup
+    values :: Map UTCTime a
+  }
+  deriving stock (Show, Eq, Generic, Functor, Foldable, Traversable)
+
+instance Applicative Timeline where
+  pure :: a -> Timeline a
+  pure a = Timeline {initialValue = a, values = mempty}
+
+  (<*>) :: forall a b. Timeline (a -> b) -> Timeline a -> Timeline b
+  fs@Timeline {initialValue = initialFunc, values = funcs} <*> xs@Timeline {initialValue, values} =
+    Timeline
+      { initialValue = initialFunc initialValue,
+        values = mergedValues
+      }
+    where
+      mergedValues :: Map UTCTime b
+      mergedValues =
+        Map.merge
+          (Map.mapMissing $ \t f -> f $ peek xs t)
+          (Map.mapMissing $ \t x -> peek fs t x)
+          (Map.zipWithMatched (const ($)))
+          funcs
+          values
+
+tshow :: Show a => a -> Text
+tshow = T.pack . show
+
+-- | Pretty-print @'Timeline' a@. It's provided so that you can investigate the
+-- value of 'Timeline' more easily. If you need to show a timeline to the end
+-- user, write your own function. We don't gurantee the result to be stable
+-- across different versions of this library.
+prettyTimeline :: forall a. Show a => Timeline a -> Text
```

Comment:
If this (and the other `pretty*`) returned `String` you could avoid a dependency on `text`.

Reply:
`text` is a [GHC boot library](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/libraries/version-history), which should be pre-installed

Reply:
I would never work with `String` for libraries which have to do any sort of real work.

Reply:
I think `String` is ok for debugging helpers like this one, especially since it is already working with `Show` instances internally.